### DISCUSSION TITLE
IDE0060 and methods "implementing" a delegate.

### DISCUSSION THREAD

**Discussion by MartyIX** [dotnet/roslyn#65273] [Q&A]
Hey, 

I wonder whether I can somehow remove the IDE0060 warning (*Remove unused parameter*) from my code:

```cs
public class TestKeeperInitSettings
{
    // [..]

    /// <inheritdoc cref="WebSocketConnectionKeeper.OnConnectedAsync"/>
    [SuppressMessage("Style", "IDE0060:Remove unused parameter", Justification = "The method signature must follow the delegate signature.")]  // <- I want to remove this line and somehow mark the method that the parameter cannot be removed.
    public Task<bool> OnConnectedCallbackAsync(StreamSubscriber streamSubscriber)
    {
        this.ConnectedEvent.Set();    
        return true;
    }
}
```

The thing is that `OnConnectedCallbackAsync`'s signature must be the same as the delegate `WebSocketConnectionKeeper.OnConnectedAsync` has. Can I somehow remove that suppression and say "this method must follow the delegate signature"?

Thank you!

**Comment by rafael-f** [comment]
You just don't want the line on your source?
You can use a suppression file:
https://learn.microsoft.com/en-us/visualstudio/code-quality/in-source-suppression-overview?view=vs-2022&tabs=csharp#suppress-violations-using-a-global-suppression-file

**Reply by MartyIX** [reply]
It's more like I would like to let the analyzer know that: "Hey, take into account that this method signature cannot be changed because it must follow the delegate signature."

**Reply by rafael-f** [reply]
I believe this is exactly what the suppression does, is this not working for you?

**Reply by MartyIX** [reply]
It works but it suppresses things. So it just moves the suppression from one place to another. But I would like to have no suppression really.

**Comment by weltkante** [comment]
> I wonder whether I can somehow remove the IDE0060 warning (Remove unused parameter) from my code

First off, this is not a warning by default, its a _suggestion_ - that is something that never should be applied blindly, but something VS is suggesting you could check for to see if it makes sense - and a lot of times these suggestions do not make sense. Turning suggestions into warnings is usually a bad idea.

> I want to remove this line and somehow mark the method that the parameter cannot be removed.

But thats what the attribute is for - marking that for the method the warning or suggestion is incorrect or irrelevant. You can be more specific than disabling it for the entire method by putting the attribute on the parameter instead.

I personally don't bother for suggestions though, since its "common sense" that suggestions should not be followed blindly, and there are tons of incorrect suggestions proposed by VS all the time, its fighting windmills to document why they should _not_ be applied.

> It works but it suppresses things. So it just moves the suppression from one place to another. But I would like to have no suppression really.

Disabling something is just another form of suppression, so "turn it off but don't suppress it" is kind of a contradiction. What technique you should use depends on how granular you want to disable it:

* for a specific syntactic element (field, parameter, property, method, class, assembly, ...): suppression attribute on the appropriate item
* for a block of source code: `#pragma warning disable/restore` commands
* for the entire project: `NoWarn` property in the project file
* for the entire solution: `NoWarn` property in `Directory.Build.props`

Remember to extend the NoWarn property, not overwrite it, so you keep suppressions contributed by other sources (like from the sdk or nuget packages):
```xml
<PropertyGroup>
  <NoWarn>IDE0060;$(NoWarn)</NoWarn>
</PropertyGroup>
```

**Reply by MartyIX** [reply]
I know I can suppress the analyzer in various ways. My question was really "Is `IDE0060` smart to handle my specific scenario?" and the answer seems to be that it's not. 

Personally, I like to have many suggestions & warnings. I'm aware that it has downsides. The `IDE0060` issue is one of them. But that doesn't mean that `IDE0060` cannot be improved by some sort of hint "hey, this follows a delegate's signature".

**Reply by weltkante** [reply]
Most of the suggestions are not "smart" and are shown innapropriately. The developer is supposed to figure out if they are appropriate and act upon them, otherwise ignore them. I don't particularly like this design of non-smart suggestions, but it is what it is.

> But that doesn't mean that IDE0060 cannot be improved

I think thats explicitly not a design goal of the suggestions, but I don't have a quote.

As far as I'm aware they don't want to cross a certain level of complexity for these suggestions (and also warnings for what its worth, but for those the bar is a little lower), for example anything that involves control flow or inspecting nonlocal state is almost always regarded as too complex.

**Reply by MartyIX** [reply]
And it makes good sense because an analyzer working with, eg, a syntax tree is cheaper than a one working with a semantic model and that in turn is cheapear than a one requiring non-trivial computations based on semantic model(s).

So yeah I guess we are on the same page.

**Comment by Youssef1313** [comment]
The analyzer should skip parameters named `_`, `_1`, `_2`, etc..

If it currently doesn't, please open an issue because I think it should ignore these.

Some background:

At some point, there was two analyzers doing the same thing, IDE0060 and CA1801. CA1801 was deprecated in favor of IDE0060. I think scenarios that were supported by CA1801 should be supported for IDE0060. See the documentation of the [deprecated rule CA1801](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1801#cause):

> This rule does not flag parameters that are named with the [discard](https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/functional/discards) symbol, for example, _, _1, and _2. This reduces warning noise on parameters that are needed for signature requirements, for example, a method used as a delegate, a parameter with special attributes, or a parameter whose value is implicitly accessed at run time by a framework but is not referenced in code.

Also note that the documentation of [IDE0060](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0060) mentions this as being a supported scenario.

**Reply by CyrusNajmabadi** [reply]
> Most of the suggestions are not "smart" and are shown innapropriately.

Please file issues for times things are shown inappropriately.
